homeassistant/components/directv/config_flow.py

```diff
@@ -3,127 +3,140 @@
-    """Validate the user input allows us to connect.
-
-    Data has the keys from DATA_SCHEMA with values provided by the user.
-    """
-    dtv = DIRECTV(data["host"], DEFAULT_PORT, determine_state=False)
-    version_info = dtv.get_version()
-
-    return {
-        "title": data["host"],
-        "host": data["host"],
-        "receiver_id": "".join(version_info["receiverId"].split()),
-    }
-
 
 class DirecTVConfigFlow(ConfigFlow, domain=DOMAIN):
     """Handle a config flow for DirecTV."""
 
     VERSION = 1
     CONNECTION_CLASS = CONN_CLASS_LOCAL_POLL
 
-    @callback
-    def _show_form(self, errors: Optional[Dict] = None) -> Dict[str, Any]:
-        """Show the form to the user."""
-        return self.async_show_form(
-            step_id="user", data_schema=DATA_SCHEMA, errors=errors or {},
-        )
-
     async def async_step_import(
-        self, user_input: Optional[Dict] = None
+        self, user_input: Optional[ConfigType] = None
     ) -> Dict[str, Any]:
-        """Handle a flow initialized by yaml file."""
+        """Handle a flow initiated by configuration file."""
         return await self.async_step_user(user_input)
 
     async def async_step_user(
-        self, user_input: Optional[Dict] = None
+        self, user_input: Optional[ConfigType] = None
     ) -> Dict[str, Any]:
-        """Handle a flow initialized by user."""
-        if not user_input:
-            return self._show_form()
-
-        errors = {}
-
-        try:
-            info = await self.hass.async_add_executor_job(validate_input, user_input)
-            user_input[CONF_HOST] = info[CONF_HOST]
-        except RequestException:
-            errors["base"] = ERROR_CANNOT_CONNECT
-            return self._show_form(errors)
-        except Exception:  # pylint: disable=broad-except
-            _LOGGER.exception("Unexpected exception")
-            return self.async_abort(reason=ERROR_UNKNOWN)
-
-        await self.async_set_unique_id(info["receiver_id"])
-        self._abort_if_unique_id_configured()
-
-        return self.async_create_entry(title=info["title"], data=user_input)
+        """Handle a flow initiated by the user."""
+        return await self._handle_config_flow(user_input)
 
     async def async_step_ssdp(
         self, discovery_info: Optional[DiscoveryInfoType] = None
     ) -> Dict[str, Any]:
-        """Handle a flow initialized by discovery."""
+        """Handle SSDP discovery."""
+        if discovery_info is None:
+            return self.async_abort(reason=ERROR_CANNOT_CONNECT)
+
         host = urlparse(discovery_info[ATTR_SSDP_LOCATION]).hostname
-        receiver_id = discovery_info[ATTR_UPNP_SERIAL][4:]  # strips off RID-
+        receiver_id = None
 
-        await self.async_set_unique_id(receiver_id)
-        self._abort_if_unique_id_configured(updates={CONF_HOST: host})
+        if discovery_info.get(ATTR_UPNP_SERIAL):
+            receiver_id = discovery_info[ATTR_UPNP_SERIAL][4:]  # strips off RID-
```

Comment:
Set it as unique ID right away, that way you know if you can abort before prompting to confirm.